Automattic/newspack-plugin
# Reduce deliverable zip size (by 70%)

### All Submissions:






### Changes proposed in this Pull Request:



_TL;DR:_ Changes here reduce the release zip file size by **70% (from 6MB to 1.8MB)**

#### JS & CSS

The JS code was bundled as separate files. See that each wizard JS file has it's own copy of `node_modules` and `assets/components`:

![image](https://user-images.githubusercontent.com/7383192/76410251-e9ab1d00-638f-11ea-9299-23db90024470.png)

Using [webpack's `SplitChunksPlugin`](https://webpack.js.org/plugins/split-chunks-plugin), the shared code is put into a `commons.js` file:

![image](https://user-images.githubusercontent.com/7383192/76410537-5d4d2a00-6390-11ea-89ac-d4de6cc40a69.png)

There's also a corresponding `commons.css` file, with reused CSS. These changes reduce the generated JS size from **15.26MB to 3.09 MB (-80%)**

#### Images 

[TinyPNG service](https://tinypng.com/) was used to reduce the size of PNG images from **3.2MB to 1.1MB (-65%)**

#### Tradeoffs

1. To use the `SplitChunksPlugin` plugin the entries in webpack config need to be strings, not arrays. I'm not sure of the reason, but the result is that the `public-path.js` snippet (which sets the correct path for non-code assets, such as images) has to be explicitly importer in all wizards' entry files.
1. Some wizards do not export their own "custom" CSS, and just use the CSS from commons.css. For this reason, the `wp_register_style` call had to be removed from those. When adding or updating a wizard, the developer will have to keep in mind whether it generates custom CSS.

### How to test the changes in this Pull Request:

1. Run `npm start` and visit all the wizards in wp-admin. Everything should work as before, no not found assets errors should be thrown (either in Network devtools console, or in the HTML, prepended to the site)

### Other information:
